### PR TITLE
Add Agda back end

### PR DESCRIPTION
**Pull request by JonasHoefer**
### Issue

[Agda Backend](https://github.com/FreeProving/free-compiler/projects/1)

### Description of the Change

- adds back end for Agda
- moves lifting from the Coq back end to the new lifted IR 

### Possible Drawbacks

Every user of the free compiler has to build the Agda dependency.

### Verification Process

We tested the new back end with a new set of tests and some case studies.

### CODE REVIEW

**Review by nbun — APPROVED** [review]

File: src/lib/FreeC/Backend/Agda/Converter/FuncDecl.hs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+-- | Converts a strongly connected component of the function dependency graph.
+--   TODO: Handle mutually recursive functions.
+convertFuncDecls
+  :: DependencyComponent IR.FuncDecl -> Converter [Agda.Declaration]
+convertFuncDecls (NonRecursive decl  ) = convertFuncDecl decl Nothing
+convertFuncDecls (Recursive    [decl]) = do
+  [decArg] <- identifyDecArgs [decl]
+  convertFuncDecl decl $ Just decArg
+convertFuncDecls (Recursive _) =
+  error "Mutual recursive functions are not supported at the moment."
```
Comment (just95) [inline src/lib/FreeC/Backend/Agda/Converter/FuncDecl.hs]: Let's use `reportFatal` here.

File: src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  NonRecursive decl   -> convertTypeDecl decl False
+  Recursive    [decl] -> convertTypeDecl decl True
+  Recursive _ ->
+    error "Mutual recursive data types are not supported at the moment."
+
+-- | Converts a data or type synonym declaration.
+--   TODO: Convert type synonyms.
+convertTypeDecl :: IR.TypeDecl -> Bool -> Converter [Agda.Declaration]
+convertTypeDecl (IR.TypeSynDecl _ _ _ _) _ =
+  error "Type synonyms are not supported at the moment."
```
Comment (just95) [inline src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs]: Use `reportFatal` here, too.

File: src/lib/FreeC/LiftedIR/Converter/Type.hs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+--   > α* = α
+liftTypeApp' :: IR.Type -> [IR.Type] -> LIR.Type
+liftTypeApp' (IR.TypeCon srcSpan name) ts =
+  LIR.TypeCon srcSpan name (map liftType' ts) False
+liftTypeApp' (IR.TypeVar srcSpan name) [] = LIR.TypeVar srcSpan name
+liftTypeApp' (IR.TypeApp _ l r       ) ts = liftTypeApp' l (r : ts)
+liftTypeApp' (IR.FuncType srcSpan l r) [] =
+  LIR.FuncType srcSpan (liftType l) (liftType r)
+liftTypeApp' _ (_ : _) =
+  error "liftTypeApp': Only type constructors can be applied!"
```
Comment (just95) [inline src/lib/FreeC/LiftedIR/Converter/Type.hs]: Due to the `KindCheckPass`, this error should not occur in practice. However, we should probably still use `reportFatal` here (which would require `liftType` to be lifted into the `Converter` monad, though).

File: src/lib/FreeC/LiftedIR/Converter/Type.hs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+--   At the moment this isn't needed because our termination checker doesn't
+--   cover cases where the decreasing argument is part of another argument.
+--   For example, a decreasing argument in the first element of a pair is not covered.
+markOutermostDecreasing :: LIR.Type -> LIR.Type
+markOutermostDecreasing (LIR.TypeCon srcSpan name ts _) =
+  LIR.TypeCon srcSpan name ts True
+markOutermostDecreasing (LIR.FreeTypeCon srcSpan t) =
+  LIR.FreeTypeCon srcSpan $ markOutermostDecreasing t
+markOutermostDecreasing _ =
+  error "Outermost argument is not necessarily unique"
```
Comment (just95) [inline src/lib/FreeC/LiftedIR/Converter/Type.hs]: This error can defiantly be triggered by using a decreasing argument pragmas. Thus `reportFatal` should be used here as well.

File: README.md [hunk 5]
```diff
@@ -61,7 +61,7 @@ This repository is structured as follows.
 
    This directory contains GitHub-related files such as issue and pull request templates as well as the configuration of the [CI pipeline][guidelines/CONTRIBUTING#the-ci-pipeline].
 
- - `./base`
+ - `./base/coq`
 
    This directory contains the Coq base library of the compiler.
    The Coq base library is a collection of Coq files that are required by the generated code.
```
Comment (just95) [inline README.md]: We should add a bullet point for the `./base/agda` library.

File: README.md [hunk 6]
```diff
@@ -155,23 +155,25 @@ The compiler has been tested with the following software versions on a Debian ba
...
 ```
 
-> **Note:** If you add or remove files from the `base` library (or any other directory that contains Coq code that you wish to compile using the script above), the automatically generated Makefile needs to be updated.
+> **Note:** If you add or remove files from the `base/coq` library (or any other directory that contains Coq code that you wish to compile using the script above), the automatically generated Makefile needs to be updated.
 > For this purpose the script provides the command line option `--recompile`.
 >
 > ```bash
-> ./tool/compile-coq.sh --recompile base
+> ./tool/compile-coq.sh --recompile base/coq
 > ```
```
Comment (just95) [inline README.md:186]: Is there anything the user needs to know about the Agda base library and how to install and use it?

File: free-compiler.cabal [hunk 7]
```diff
@@ -32,29 +32,29 @@ homepage:            https://github.com/FreeProving/free-compiler
...
+                   , base/coq/Test/*.v
+                     -- base/coq/**/*.vo
+                   , base/coq/Prelude.vo
+                   , base/coq/Prelude/*.vo
+                   , base/coq/Free.vo
+                   , base/coq/Free/*.vo
+                   , base/coq/Free/Instance/*.vo
+                   , base/coq/Free/Tactic/*.vo
+                   , base/coq/Free/Util/*.vo
+                   , base/coq/Test/*.vo
```
Comment (just95) [inline free-compiler.cabal]: We should maybe copy the Agda base library to the compiler's installation directory as well.

File: .github/workflows/ci-pipeline.yml [hunk 8]
```diff
@@ -402,7 +432,7 @@ jobs:
         ./freec                                          \
           --transform-pattern-matching                   \
           --dump-transformed-modules example/transformed \
-          -b ./base                                      \
+          -b ./base/coq                                  \
```
Comment (just95) [inline .github/workflows/ci-pipeline.yml]: It's a bit weird that the user has to point to the Coq base library even if they want to compile to Agda. The `*.toml` files should be moved to `./base` or there should be another set of `*.toml` files in `./base/agda`.

**Review by just95 — CHANGES_REQUESTED** [review]
In this review I've tried to focus on the user's perspective. I've added some comments for how to improve the documentation and error messages. Additionally, we should mention the addition of an Agda backend in the `CHANGELOG.md`.

**Review by just95 — APPROVED** [review]

**Pull request merged**